### PR TITLE
Refactor: introduce general `operate` commands to reduce duplicate code

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description

As title, introduce general operate commands to reduce duplicate code

And by the way, make `format filesize` command more flexible, so it can accept input from pipeline, and map input row.

### format filesize
The changes idea comes from: https://discord.com/channels/601130461678272522/614593951969574961/1021993690631241788

By default, `format filesize` is overkill, because we always need to create a table first.

### refactor
After hacking into nushell code, I found there are too much function called `operate`, I think we can pick them out as a general `operate` function, and refactor `bytes`, and `str` relative commands.

### format filesize
`format filesize`'s input handling behavior is similar to str subcommands, which accept a simple value, a table, a record, or something else as input pipeline data.

### refactor
nushell have less code :-)

## About the change
This pr introduces a function named `operate`(which exists in many places around the code) to handle input as *map over table* approach.

# Tests

Make sure you've done the following:

- [X] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [X] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [X] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [X] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [X] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [X] `cargo test --workspace --features=extra` to check that all the tests pass

# Documentation

- [X] If your PR touches a user-facing nushell feature then make sure that there is an entry in the documentation (https://github.com/nushell/nushell.github.io) for the feature, and update it if necessary.

### CODE REVIEW

**Comment by WindSoilder** [comment]
Mark as draft for now, I'm not sure the `operate` function is good to introduce, causes we have `pipeline_data.map` function which is usable

**Comment by sholderbach** [comment]
I am not sure yet what the `trait Argument` should do or if the naming fully resonates with me. Would this be your proposed way to make calling with argument or pipeline generic?

**Comment by WindSoilder** [comment]
> Would this be your proposed way to make calling with argument or pipeline generic?

Yeah, it's borrowed from 
https://github.com/nushell/nushell/blob/ebca840d91e81c070800e39cea07325d25b4fc54/crates/nu-command/src/bytes/mod.rs#L32-L34

Which is also borrowed from:
https://github.com/nushell/nushell/blob/ebca840d91e81c070800e39cea07325d25b4fc54/crates/nu-command/src/path/mod.rs#L32-L34

File: crates/nu-command/tests/commands/format.rs [hunk 1]
```diff
@@ -80,7 +80,7 @@ fn format_filesize_works_with_nonempty_files() {
 
             let actual = nu!(
                 cwd: dirs.test(),
-                "ls sample.toml | format filesize size B | get size | first"
+                "ls sample.toml | format filesize B size | get size | first"
```
Comment (jntrnr) [inline crates/nu-command/tests/commands/format.rs:83]: Does this mean that we'll be able to pass multiple columns?

**Comment by sholderbach** [comment]
> Yeah, it's borrowed from
> 
> https://github.com/nushell/nushell/blob/ebca840d91e81c070800e39cea07325d25b4fc54/crates/nu-command/src/bytes/mod.rs#L32-L34
> 
> Which is also borrowed from:
> 
> https://github.com/nushell/nushell/blob/ebca840d91e81c070800e39cea07325d25b4fc54/crates/nu-command/src/path/mod.rs#L32-L34

Would this make sense to be part of a larger refactor? I am a bit wary of introducing a pattern for just one command, if this just applies to one command I would favor a simpler more readable solution. (and better naming to describe what it does `Arguments` is pretty specific to something we already kind of have in the `Signature`)

Unrelated: Would you maybe provide a before and after comparison what this change allows on this particular command? I don't know if I fully grok it from the tests

**Comment by WindSoilder** [comment]
> Would this make sense to be part of a larger refactor?

Yeah, sorry I didn't mention it in the pr.  Since many command shares the same pattern like this, I'd like to introduce it in generate `operate` function (provide in this pr).

Then we'll have oppounity to simplify bytes and str commands, and unify input handling behavior in general based on the new `operate` function and `Argument` trait.  So we don't need to define separate `operation` function anymore.  I just think it can reduce our code.

This pr is just provide an idea, and `format filesize` is an example of how to use `operate` function and `Argument` trait

File: crates/nu-command/tests/commands/format.rs (hunk 1 above)
Comment (WindSoilder) [reply crates/nu-command/tests/commands/format.rs:83]: Yeah, like other `str` commands

**Comment by WindSoilder** [comment]
I'll mark as draft, make a little larger refactor on `bytes`, `str`, `path` commands to see how's going on.

**Comment by WindSoilder** [comment]
Ok, currently I just refactor `bytes` and `str` commands, It's a relative large change.

But doesn't change commands' behavior(I havent change test code's checking logic, to make sure I don't break anything) would you please have a review? @sholderbach @jntrnr

**Comment by sholderbach** [comment]
From the cursory look, I am happy with the changes and didn't notice things that are particularly risky for breakage. But I can give it a deeper look pulling in the test coverage.

In general refactors that separate the actual unique logic from the repetitive boiler plate should make future engine-r changes easier (or stuff like the I/O type proposal from @dandavison in #6796) so I don't think this is blocking and I think the first part of engine-r will probably be the new parser independent of the type or engine side.
And having smaller command files is just nicer if you have to work on them in the future!

**Comment by dandavison** [comment]
This looks like a great clean up! I think it would have become untenable eventually to have that stuff repeated in all commands. 

Am I understanding correctly that there used to be something called "column paths" and now we have "cell paths"? It would be good to rename `column_paths` in the code. Perhaps this PR would be a good opportunity to do so?

```
let column_paths: Vec<CellPath> = call.rest(engine_state, stack, 1)?;
```
```
    fn take_column_paths(&mut self) -> Option<Vec<CellPath>> {
        self.column_paths.take()
    }
```

**Comment by WindSoilder** [comment]
> Am I understanding correctly that there used to be something called "column paths" and now we have "cell paths"? It would be good to rename column_paths in the code. Perhaps this PR would be a good opportunity to do so?

Oh, that's a good catch!  I remember jt have talked about that before, I think it's good to rename from column path to cell path for relative commands, and see how is going

**Comment by fdncred** [comment]
Great work @WindSoilder. So, happy you're helping us!

**Comment by sholderbach** [comment]
I looked over everything after running `cargo llvm-cov` with [my old script](https://github.com/nushell/nu_scripts/blob/main/make_release/get_coverage.nu) and all the commands are covered 80-95%. Thanks to common source of truth for the cellpath handling the numbers should have increased from the state before as not every command might have a test specifically for that. Even more benefits from your awesome work, @WindSoilder!

So I think we can vouch for the fact that no breaking change was introduced apart from the correction of the argument order in `format filesize`. So that would be something for the [release notes](https://github.com/nushell/nushell.github.io/blob/release-notes-0.71/blog/2022-11-08-nushell-0.71.md)

Happy to land this

**Review by sholderbach — APPROVED** [review]

**Comment by fdncred** [comment]
same.  good work @WindSoilder. thanks!
